### PR TITLE
Feat: Check freshness for mixed external & sqlmesh models

### PR DESCRIPTION
**Pull request by VaggelisD**
Follow up PR on model freshness ([Previous PR](https://github.com/TobikoData/sqlmesh/pull/5277))

This PR enables mixed SQLMesh & external (source) model freshness, e.g:

```
A: External model 
B: SELECT * FROM A
C: SELECT * FROM A, B
``` 

Model `C` will be evaluated if its upstream models are fresh, which is determined by:
- `C`'s `last_altered_ts` compared to external model's (`A`) `INFORMATION_SCHEMA` metadata
- The existence (or not) of parent intervals for SQLMesh models (`B`), signalling if any model will be evaluated as part of this run

A new freshness-specific test file is added to break up the original test in various scenarios.

### CODE REVIEW

File: .circleci/continue_config.yml [hunk 1]
```diff
@@ -313,10 +313,10 @@ workflows:
...
                 - fabric
                 - gcp-postgres
-          filters:
-            branches:
-              only:
-                - main
+          # filters:
+          #   branches:
+          #     only:
+          #       - main
```
Comment (VaggelisD) [inline .circleci/continue_config.yml]: TODO: Reinstate before merging

File: sqlmesh/core/signal.py [hunk 2]
```diff
@@ -42,7 +44,17 @@ class signal(registry_decorator):
 
 
 @signal()
-def freshness(batch: DatetimeRanges, snapshot: Snapshot, context: ExecutionContext) -> bool:
+def freshness(
+    batch: DatetimeRanges,
+    snapshot: Snapshot,
+    context: ExecutionContext,
+    parent_intervals: t.Optional[t.List[Intervals]] = None,
```
Comment (izeigerman) [inline sqlmesh/core/signal.py]: Can we somehow make this a part of `ExecutionContext` rather than adding another argument?

File: sqlmesh/core/signal.py [hunk 3]
```diff
@@ -54,24 +66,38 @@ def freshness(batch: DatetimeRanges, snapshot: Snapshot, context: ExecutionConte
...
     )
+
     if not last_altered_ts:
         return True
 
     parent_snapshots = {context.snapshots[p.name] for p in snapshot.parents}
-    if len(parent_snapshots) != len(snapshot.node.depends_on) or not all(
-        p.is_external for p in parent_snapshots
-    ):
+    if len(parent_snapshots) != len(snapshot.node.depends_on):
```
Comment (izeigerman) [inline sqlmesh/core/signal.py]: I'm still confused. If the 2 sets don't match that means the model references external tables. Why don't we treat them just like external models?

File: sqlmesh/core/signal.py (hunk 2 above)
Comment (VaggelisD) [reply sqlmesh/core/signal.py]: Agreed, was hesitant to keep loading `ExecutionContext` but it's a much cleaner approach.

File: sqlmesh/core/signal.py (hunk 3 above)
Comment (VaggelisD) [reply sqlmesh/core/signal.py]: Removed this condition in the latest commit. To account for any edge cases e.g an external model not being present in the `INFORMATION_SCHEMA`, I added a raise condition (and a test for it) instead.

File: sqlmesh/core/signal.py [hunk 4]
```diff
@@ -54,24 +64,35 @@ def freshness(batch: DatetimeRanges, snapshot: Snapshot, context: ExecutionConte
...
     parent_snapshots = {context.snapshots[p.name] for p in snapshot.parents}
-    if len(parent_snapshots) != len(snapshot.node.depends_on) or not all(
-        p.is_external for p in parent_snapshots
-    ):
-        # The mismatch can happen if e.g an external model is not registered in the project
+
+    upstream_parent_snapshots = {p for p in parent_snapshots if not p.is_external}
+    external_parents = snapshot.node.depends_on - {p.name for p in upstream_parent_snapshots}
+
+    if context.parent_intervals:
```
Comment (izeigerman) [inline sqlmesh/core/signal.py:76]: We should exclude external models from this check. There could be intervals for external models to, for example, run audits, but it doesn't mean that the table is actually fresh.

File: sqlmesh/core/scheduler.py [hunk 5]
```diff
@@ -364,13 +364,21 @@ def batch_intervals(
 
             adapter = self.snapshot_evaluator.get_adapter(snapshot.model_gateway)
 
+            parent_intervals = []
+            for parent in snapshot.parents:
+                if parent.snapshot_id not in snapshot_intervals:
+                    continue
+                _, p_intervals = snapshot_intervals[parent.snapshot_id]
+                parent_intervals.append(p_intervals)
```
Comment (izeigerman) [inline sqlmesh/core/scheduler.py]: I don't think this is the right place to construct this. The missing intervals will change as we apply signal checks to snapshots. If you set them once here, the parent signals will not be reflected.

File: sqlmesh/core/signal.py (hunk 4 above)
Comment (VaggelisD) [reply sqlmesh/core/signal.py:76]: Right. I guess I can exclude them at the time of `parent_intervals` construction, since that's a list so we've lost the snapshot mapping of each `Interval`.

File: sqlmesh/core/scheduler.py (hunk 5 above)
Comment (VaggelisD) [reply sqlmesh/core/scheduler.py]: Good call.

 I guess our best choice would be to check `snapshot_batches` then (same loop) which applies the signal to filter out the unready intervals. 

So, for each snapshot in the DAG, its `ExecutionContext::parent_intervals` list will be extended by the `snapshot_batches[parent]`, provided the `parent` snapshot is NOT external.

Comment (izeigerman) [reply sqlmesh/core/scheduler.py]: yes, this sounds correct, assuming I understood the idea correctly

Comment (VaggelisD) [reply sqlmesh/core/scheduler.py]: Implemented in https://github.com/TobikoData/sqlmesh/pull/5499/commits/9a8af2e6fe9bacf1c97178681a785804b77230a8

File: sqlmesh/core/signal.py (hunk 4 above)
Comment (VaggelisD) [reply sqlmesh/core/signal.py:76]: Implemented in https://github.com/TobikoData/sqlmesh/pull/5499/commits/9a8af2e6fe9bacf1c97178681a785804b77230a8, external snapshots are excluded from `parent_intervals`.

File: sqlmesh/core/scheduler.py [hunk 6]
```diff
@@ -364,13 +364,22 @@ def batch_intervals(
 
             adapter = self.snapshot_evaluator.get_adapter(snapshot.model_gateway)
 
+            parent_intervals: Intervals = []
+            for parent_id in snapshot.parents:
+                parent_snapshot, _ = snapshot_intervals.get(parent_id, (None, []))
```
Comment (izeigerman) [inline sqlmesh/core/scheduler.py:369]: Are you sure that `snapshot_intervals` reflects updated intervals after signals were applied to parents. If I'm reading the code correctly, `snapshot_intervals` remain unchanged and only `snapshot_batches` reflect signal outcomes.

File: sqlmesh/core/signal.py [hunk 7]
```diff
@@ -54,24 +64,35 @@ def freshness(batch: DatetimeRanges, snapshot: Snapshot, context: ExecutionConte
...
         return True
 
     parent_snapshots = {context.snapshots[p.name] for p in snapshot.parents}
-    if len(parent_snapshots) != len(snapshot.node.depends_on) or not all(
-        p.is_external for p in parent_snapshots
-    ):
-        # The mismatch can happen if e.g an external model is not registered in the project
+
+    upstream_parent_snapshots = {p for p in parent_snapshots if not p.is_external}
+    external_parents = snapshot.node.depends_on - {p.name for p in upstream_parent_snapshots}
```
Comment (izeigerman) [inline sqlmesh/core/signal.py:74]: Sorry man, I'm still confused. `external_parents` contains external tables that were **not** registered as external models. Both `upstream_parent_snapshots` **and** `context.parent_intervals` exclude registered external models. So at which point do we check the last modified timestamp for tables that are registered external models?

Comment (izeigerman) [reply sqlmesh/core/signal.py:74]: I believe `external_models` should include both:
```
snapshot.node.depends_on - {p.name for p in upstream_parent_snapshots}
```
AND
```
{p.name for p in parent_snapshots if p.is_external}
```
am I missing something?

File: sqlmesh/core/scheduler.py (hunk 6 above)
Comment (VaggelisD) [reply sqlmesh/core/scheduler.py:369]: We only use `snapshot_interval` here to retrieve the parent snapshots. 

Since the latest commit, the new `parent_intervals` are only based on `snapshot_batches`, see L373. I'll add a test that verifies this given a parent model that has a custom signal suddenly evaluating to `False`.

File: sqlmesh/core/signal.py (hunk 7 above)
Comment (VaggelisD) [reply sqlmesh/core/signal.py:74]: My thinking was the following: For every model, its parents (i.e the entirety of `model.depends_on` set) can be:
1. SQLMesh models
2. Registered external models (we have a `Snapshot` for them)
3. Unregistered external models

So, given all of the parent **snapshots** i.e (1) + (2):
- `upstream_parent_snapshots` computes (1)
- `external_parents` gathers both (2) and (3) by doing a set difference on `model.depends_on - (1)`

I guess the gotcha here is that we don't really use `upstream_parent_snapshots` elsewhere, but I wanted to break the computation of `external_parents` in steps.

So, the actual freshness computation comes from:
- The `context.parent_intervals`, which indicates whether **any** model from (1) has **any** ready intervals
- The model's `last_altered_ts` comparison with the `external_models` metadata i.e (2) + (3)


Does this make more sense? I hope I'm not butchering any set operation here

File: sqlmesh/core/scheduler.py (hunk 6 above)
Comment (VaggelisD) [reply sqlmesh/core/scheduler.py:369]: I added a new test here that should exercise this scenario:

https://github.com/TobikoData/sqlmesh/commit/d4e3fbe91d4c58e06bca5e3ec869708f78184a87#diff-4dff5ecd31963149685166599deae8171b2cda7a1e8bcc99476a173b04925c33

File: sqlmesh/core/signal.py (hunk 7 above)
Comment (VaggelisD) [reply sqlmesh/core/signal.py:74]: Added a new test to validate this behavior: https://github.com/TobikoData/sqlmesh/pull/5499/commits/de4630620a51c7ba115f102d34bc54e486a01925
